components/feral-connectd/cdp.go

```
@@ -0,0 +1,154 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"strconv"
+	"sync"
+
+	"github.com/gorilla/websocket"
+)
+
+// CDP Methods
+const (
+	NavigateMethod = "Page.navigate"
+	EvaluateMethod = "Runtime.evaluate"
+)
+
+type CDPClient struct {
+	conn   *websocket.Conn
+	reqID  int
+	config *Config
+	mu     sync.Mutex
+	done   chan struct{}
+}
+
+// NewCDPClient creates a new CDP client
+func NewCDPClient(config *Config) *CDPClient {
+	return &CDPClient{
+		config: config,
+		reqID:  0,
+		done:   make(chan struct{}),
+	}
+}
+
+// InitCDP fetches WS endpoint and dials Chromium
+func (c *CDPClient) InitCDP(ctx context.Context) error {
+	// Fetch JSON with websocket debugger URL
+	resp, err := http.Get(c.config.CDPHost + ":" + strconv.Itoa(c.config.CDPPort) + "/json")
+	if err != nil {
+		return fmt.Errorf("failed to fetch debug targets: %w", err)
+	}
+
+	defer resp.Body.Close()
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return fmt.Errorf("failed to read targets: %w", err)
+	}
+
+	var targets []struct {
+		Type                 string `json:"type"`
+		WebSocketDebuggerURL string `json:"webSocketDebuggerUrl"`
+	}
+	if err := json.Unmarshal(body, &targets); err != nil {
+		return fmt.Errorf("invalid targets format: %w", err)
+	}
+
+	// Connect to CDP websocket
+	for _, t := range targets {
+		if t.Type == "page" {
```

In theory, there could be a chance that we can see multiple `page` if there is one more tab opened or even not `page` if the chromium cannot be launched before. Even though we should avoid it from chromium launching, we should still handle it well here. My suggestion is to loop through all the targets and collect all `page`. If the page array length is not 1, we should return the error. Otherwise, just pick the underlying page and process further more. By that way, we also can make the code clearer and remove the obsolete indents.